File: frontend/src/utils/cryptoIcons.ts
/** Brand colors + glyph/CDN helpers for cryptocurrency UI badges. */

const BRAND: Record<string, { color: string; glyph: string; slug: string }> = {
  BTC: { color: '#F7931A', glyph: '₿', slug: 'btc' },
  ETH: { color: '#627EEA', glyph: 'Ξ', slug: 'eth' },
  USDT: { color: '#26A17B', glyph: '₮', slug: 'usdt' },
  USDC: { color: '#2775CA', glyph: '$', slug: 'usdc' },
  BNB: { color: '#F3BA2F', glyph: '◆', slug: 'bnb' },
  LTC: { color: '#345D9D', glyph: 'Ł', slug: 'ltc' },
  TRX: { color: '#FF0013', glyph: 'T', slug: 'trx' },
  XRP: { color: '#23292F', glyph: '✕', slug: 'xrp' },
  SOL: { color: '#9945FF', glyph: '◎', slug: 'sol' },
  DOGE: { color: '#C2A633', glyph: 'Ð', slug: 'doge' },
  MATIC: { color: '#8247E5', glyph: 'M', slug: 'matic' },
  POL: { color: '#8247E5', glyph: 'P', slug: 'matic' },
  AVAX: { color: '#E84142', glyph: 'A', slug: 'avax' },
  ADA: { color: '#0033AD', glyph: '₳', slug: 'ada' },
  DOT: { color: '#E6007A', glyph: '●', slug: 'dot' },
  LINK: { color: '#2A5ADA', glyph: '⬡', slug: 'link' },
  SHIB: { color: '#FFA409', glyph: 'S', slug: 'shib' },
  DAI: { color: '#F5AC37', glyph: '◈', slug: 'dai' },
  BUSD: { color: '#F0B90B', glyph: 'B', slug: 'busd' },
  TON: { color: '#0098EA', glyph: '◆', slug: 'ton' },
}

/** Base ticker from composite symbols like USDT_TRC20 → USDT */
export function cryptoBase(symbol?: string | null): string {
  if (!symbol) return ''
  return symbol.split(/[_\s/-]/)[0].toUpperCase()
}

export function cryptoBrand(symbol?: string | null) {
  const base = cryptoBase(symbol)
  return (
    BRAND[base] || {
      color: '#64748B',
      glyph: base.slice(0, 1) || '?',
      slug: base.toLowerCase() || 'generic',
    }
  )
}

/** Official-style SVG from cryptocurrency-icons package (jsDelivr). */
export function cryptoIconUrl(symbol?: string | null): string {
  const { slug } = cryptoBrand(symbol)
  return `https://cdn.jsdelivr.net/npm/cryptocurrency-icons@0.18.1/svg/color/${slug}.svg`
}

export function cryptoGlyph(symbol?: string | null, fallbackIcon?: string | null): string {
  if (fallbackIcon && fallbackIcon.length <= 3 && !fallbackIcon.includes(' ')) {
    return fallbackIcon
  }
  return cryptoBrand(symbol).glyph
}

export function cryptoColor(symbol?: string | null): string {
  return cryptoBrand(symbol).color
}

File: core/templatetags/crypto_icons.py
"""Render offline crypto brand badges in Django templates."""
from django import template
from django.utils.html import format_html
from django.utils.safestring import mark_safe

register = template.Library()

BRAND = {
    'BTC': ('#F7931A', '₿'),
    'ETH': ('#627EEA', 'Ξ'),
    'USDT': ('#26A17B', '₮'),
    'USDC': ('#2775CA', '$'),
    'BNB': ('#F3BA2F', 'B'),
    'LTC': ('#345D9D', 'Ł'),
    'TRX': ('#FF0013', 'T'),
    'XRP': ('#23292F', 'X'),
    'SOL': ('#9945FF', 'S'),
    'DOGE': ('#C2A633', 'Ð'),
    'MATIC': ('#8247E5', 'M'),
    'POL': ('#8247E5', 'P'),
    'AVAX': ('#E84142', 'A'),
    'ADA': ('#0033AD', '₳'),
    'DOT': ('#E6007A', '●'),
    'LINK': ('#2A5ADA', '⬡'),
    'DAI': ('#F5AC37', '◈'),
    'BUSD': ('#F0B90B', 'B'),
    'TON': ('#0098EA', '◆'),
    'USD': ('#16A34A', '$'),
    'EUR': ('#2563EB', '€'),
    'GBP': ('#7C3AED', '£'),
    'UGX': ('#DC2626', 'U'),
}


def _base(symbol: str) -> str:
    if not symbol:
        return ''
    upper = str(symbol).strip().upper()
    for key in sorted(BRAND.keys(), key=len, reverse=True):
        if upper == key or upper.startswith(f'{key}_') or upper.startswith(f'{key}-') or upper.startswith(f'{key} '):
            return key
    return upper.split('_')[0].split('-')[0].split()[0]


def _brand(symbol: str):
    base = _base(symbol)
    if base in BRAND:
        return base, BRAND[base][0], BRAND[base][1]
    letter = (base[:1] or '?').upper()
    return base, '#64748B', letter


@register.simple_tag
def crypto_icon(symbol, size='sm'):
    """Colored circular crypto badge (offline). size: xs|sm|md|lg"""
    if not symbol:
        return ''
    base, color, letter = _brand(str(symbol))
    size = size if size in ('xs', 'sm', 'md', 'lg') else 'sm'
    return format_html(
        '<span class="ci-crypto-icon ci-crypto-{}" style="background:{}" title="{}" aria-label="{}">'
        '<span class="ci-crypto-letter">{}</span></span>',
        size,
        color,
        str(symbol),
        str(symbol),
        letter,
    )


@register.simple_tag
def crypto_label(symbol, size='sm', network=''):
    """Icon + symbol text row."""
    if not symbol:
        return mark_safe('—')
    icon = crypto_icon(symbol, size=size)
    if network:
        return format_html(
            '<span class="ci-crypto-label">{} <span class="ci-crypto-meta">'
            '<span class="ci-crypto-sym">{}</span>'
            '<span class="ci-crypto-net">{}</span></span></span>',
            icon,
            str(symbol),
            str(network),
        )
    return format_html(
        '<span class="ci-crypto-label">{} <span class="ci-crypto-sym">{}</span></span>',
        icon,
        str(symbol),
    )
